doc/developer/design/20220228_kafka_headers.md

```diff
@@ -0,0 +1,104 @@
+# Kafka Headers
+
+## Summary
+
+Support the `headers` section of kafka messages in sources. This is a `string` -> `bytes` map of arbitrary
+data. The purpose of this design document is to describe a multi-phase implementation of this
+
+## Goals
+
+1. Begin support with basic support for unstructured data from headers that can be cast/converted later
+in the sql layer
+2. Extend support to include typed, named headers
+
+## Non-Goals
+
+It is a non-goal to support
+[Kafka Connect `Header`'s](https://kafka.apache.org/20/javadoc/index.html?org/apache/kafka/connect/header/Header.html),
+which are, as far as I can tell, _dynamically-schema'd_ headers built on kafka headers. This is similar to the
+dynamic schema's supported by Kafka Connect.
+
+
+## Description
+
+
+### Phase 1
+
+Add an `INCLUDE HEADERS` syntax that allows users to declare that they want a `headers` column for their source,
+as provided in kafka sources. This will provide a map from text to bytes (`bytea`).
+Users are expected to cast these bytes values to the types they want in a later view.
+This option will only be available for `ENVELOPE NONE` and `ENVELOPE UPSERT`.
+
+### Phase 2
+As an extension to phase 1, allows users to pre-declare their expected types in the source definition, simplifying their usage.
+
+## Proposed syntax changes
+
+In phase 1, kafka source creation will be extended so that `INCLUDE` has a new `HEADERS` option, so the full
+syntax becomes:
+
+```
+INCLUDE (
+    KEY (AS <name>)? |
+    TIMESTAMP (AS <name>)? |
+    OFFSET (AS <name>)? |
+    PARTITION (AS <name>)? |
+    TOPIC (AS <name>)? |
+    HEADERS \(headername...\)?
+)+
+```
+as an example:
+```sql
+CREATE MATERIALIZED SOURCE avroavro
+FROM KAFKA BROKER '...' TOPI '...`
+FORMAT ...
+INCLUDE HEADERS
+ENVELOPE ...
+```
+
+This will add a new column to the row from the kafka source that is of type
+`map[text=>bytea]` with the name `headers`, if it doesn't clash
+with an existing column.
+
+In phase 2, the syntax will be
+```
+INCLUDE (
+    KEY (AS <name>)? |
+    TIMESTAMP (AS <name>)? |
+    OFFSET (AS <name>)? |
+    PARTITION (AS <name>)? |
+    TOPIC (AS <name>)? |
+    HEADERS \(headername( type)...\)?
+)+
+```
+where `type` will override the type from `bytes` to the given one.
```

Comment:
Eh, I'm not convinced we ever need to do this! We're trying to push things in the direction of "write these sorts of transformations in views downstream of your source", to de-complicate the `CREATE SOURCE` syntax.

Reply:
will add a note!